src/script/interpreter.cpp

```diff
@@ -1681,6 +1752,205 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                        {
+                            if (!inp.assetIssuance.IsNull()) {
+                                valtype vchAssetBlindingNonce, vchAssetEntropy;
+                                pushvalue(stack, inp.assetIssuance.nInflationKeys);
+                                pushvalue(stack, inp.assetIssuance.nAmount);
+                                // Next push Asset entropy
+                                vchAssetEntropy.insert(vchAssetEntropy.begin(), inp.assetIssuance.assetEntropy.begin(), inp.assetIssuance.assetEntropy.end());
+                                stack.push_back(vchAssetEntropy);
+                                // Finally push blinding nonce
+                                // By pushing the this order, we make sure that the stack top is empty
+                                // iff there is no issuance. Note that nInflationKeys can be null and can push false
+                                vchAssetBlindingNonce.insert(vchAssetBlindingNonce.begin(), inp.assetIssuance.assetBlindingNonce.begin(), inp.assetIssuance.assetBlindingNonce.end());
+                                stack.push_back(vchAssetBlindingNonce);
+                            } else { // No issuance
+                                stack.push_back(vchFalse);
+                            }
+                            break;
+                        }
+                        default: assert(!"invalid opcode"); break;
+                    }
+                }
+                break;
+
+                case OP_PUSHCURRENTINPUTINDEX:
+                {
+                    // OP_PUSHCURRENTINPUTINDEX is available post tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+
+                    if (checker.GetnIn() > 0x7fffffff)
+                        return set_error(serror, SCRIPT_ERR_UNKNOWN_ERROR);
+                    CScriptNum currIdx = CScriptNum((int64_t) checker.GetnIn());
+                    stack.push_back(currIdx.getvch());
+                }
+                break;
+
+                case OP_INSPECTOUTPUTASSET:
+                case OP_INSPECTOUTPUTVALUE:
+                case OP_INSPECTOUTPUTNONCE:
+                case OP_INSPECTOUTPUTSCRIPTPUBKEY:
+                {
+                    // Output instropsection codes only available post tapscript is available post tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+
+                    if (stack.size() < 1)
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+
+                    int idx = CScriptNum(stacktop(-1), fRequireMinimal).getint();
+                    popstack(stack);
+
+                    const PrecomputedTransactionData *cache = checker.GetPrecomputedTransactionData();
+                    auto outs = checker.GetTxvOut();
```

Comment:
This pointer could be null in exotic SignatureChecker contexts.  I don't think we can assume it is non-null.

Reply:
There are no exotic SignatureChecker contexts except in testing envionments.

Reply:
Actually, in Elements there is, because we have a script interpreter in blockheaders :).

Reply:
https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki

Reply:
Regardless you cannot have a post-condition for BaseSignatureChecker be that the pointer is non-null and then implement it with "return nullptr".  Either assert false or make it pure virtual, or deal with the fact that the post-condition is that the pointer can be null.

Requiring a whole program analysis to understand the correctness of this is not acceptable.

Reply:
Given that we have to live with BaseSignatureChecker's inheritance hierarchy mess for the time being, the way I'd like to see this improved is to have

1. have these pointer returning methods return references instead of pointers.
2. assert false for most or all of these default methods (instead of returning null pointers).

Reply:
The alternative is to explicitly handle these null pointers in EvalScript, by calling seterror and returning false.

Reply:
This is what is currently implemted. I think the correct solution is to return a type `Option<reference>`, but I suspect that will leak throughout the codebase.

Reply:
Okay after looking into (legacy) `DataFromTransaction`, I now believe what we have written is, more or less, the best we can do.

However we need some sort of TODO item here to assert false depending on the MissingDataBehavior setting after it is introduced in https://github.com/bitcoin/bitcoin/pull/21330.